urlaa051fbd3d_WheezePuppet_collegeDiversity_tgz-pJ8-edu_umw_cpsc_collegediversity_SimJ8/src/edu/umw/cpsc/collegediversity/Student.java

```diff
@@ -0,0 +1,439 @@
+    }
+
+    /**
+     * Return this student's GPA (on a 4.0 scale).
+     */
+    public double getGpa(){
+        return gpa;
+    }
+
+    /**
+     * Return this student's race.
+     */
+    public Race getRace(){
+        return race;
+    }
+
+    private void setId(int Id){
+        id=Id;
+    }
+
+    private void setGrade(int grade){
+        this.grade=grade;
+    }
+
+    private void setGpa(double gpa){
+        this.gpa=gpa;
+    }
+
+    private void setRace(Race race){
+        this.race=race;
+    }
+
+    /**
+     * Assign this student to a new dorm room. This method <i>will</i>
+     * inform the corresponding {@link Room} object that it has the new
+     * occupant, and if this student previously had a room, that Room
+     * object will be told that it no longer has the occupant.
+     */
+    public void setRoom(Room room){
+        if (room == null) {
+            //System.out.println("** Student #" + id + " leaving their room.");
+        }
+
+        if (dormRoom != null) {
+            dormRoom.removeOccupant(this);
+        }
+        dormRoom = room;
+        dormRoom.addOccupant(this);
+    }
+
+    // Add attribute number "i" to this student. Should only be done at
+    // instantiation time, from FreshmanFactory.
+    void addAttribute(int i){
+        attributes.add(i);
+    }
+
+    private static final double SMALL = 1/1000.0;
+
+    private boolean isSeptember(double time) {
+        if (Math.abs(time - Math.floor(time) - 1/12.0) < SMALL) {
+            return true;
+        }
+        return false;
+    }
+
+    private boolean isMay(double time) {
+        if (Math.abs(time - Math.floor(time) - 9/12.0) < SMALL) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Perform monthly activities for this student. In particular:
+     * <ol>
+     * <li>On Sept. 1st, assign a dorm room, classes (TODO), and, if 
+     * freshman (TODO), an orientation group.</li>
+     * <li>On Oct., Nov., Dec., Jan., Feb., Mar., Apr. 1st, possibly
+     * encounter (and possibly form connections with) other students.</li>
+     * <li>On May 1st., increment the academic year, and possibly drop out
+     * or graduate. (Note that the {@link StatsPrinter} will run on April
+     * 30th, just before this promotion activity, to dump out stats for the
+     * just-now-ending academic year's students.)</li>
+     * <li>On Jun., Jul., Aug. 1st, do nothing (when performing year-end
+     * activities on May 1st, this object will be scheduled for four months
+     * away, on Sept. 1st, instead of one month away).</li>
+     * </ol>
+     */
+    public void step(SimState state){
+        if (isSeptember(state.schedule.getTime())) {
+            performYearBeginningActivities();
+        } else if (isMay(state.schedule.getTime())) {
+            performYearEndingActivities();
+        } else {
+            performMonthlyActivities();
+        }
+    }
+
+    private void performYearBeginningActivities() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 230-238 into a separate method.